lib/spack/spack/binary_distribution.py

```diff
@@ -1597,57 +1649,87 @@ def needs_rebuild(spec, mirror_url, rebuild_on_errors=False):
         pkg_name, pkg_version, pkg_hash, pkg_full_hash))
     tty.debug(spec.tree())
 
-    # Try to retrieve the .spec.yaml directly, based on the known
+    # Try to retrieve the specfile directly, based on the known
     # format of the name, in order to determine if the package
     # needs to be rebuilt.
     cache_prefix = build_cache_prefix(mirror_url)
-    spec_yaml_file_name = tarball_name(spec, '.spec.yaml')
-    file_path = os.path.join(cache_prefix, spec_yaml_file_name)
+    specfile_is_json = True
+    specfile_name = tarball_name(spec, '.spec.json')
+    deprecated_specfile_name = tarball_name(spec, '.spec.yaml')
+    specfile_path = os.path.join(cache_prefix, specfile_name)
+    deprecated_specfile_path = os.path.join(cache_prefix,
+                                            deprecated_specfile_name)
 
     result_of_error = 'Package ({0}) will {1}be rebuilt'.format(
         spec.short_spec, '' if rebuild_on_errors else 'not ')
 
     try:
-        _, _, yaml_file = web_util.read_from_url(file_path)
-        yaml_contents = codecs.getreader('utf-8')(yaml_file).read()
+        if web_util.url_exists(deprecated_specfile_path):
+            _, _, spec_file = web_util.read_from_url(deprecated_specfile_path)
+            specfile_is_json = False
+        elif web_util.url_exists(specfile_path):
+            _, _, spec_file = web_util.read_from_url(specfile_path)
+        else:
+            raise URLError('')
+        spec_file_contents = codecs.getreader('utf-8')(spec_file).read()
     except (URLError, web_util.SpackWebError) as url_err:
         err_msg = [
             'Unable to determine whether {0} needs rebuilding,',
             ' caught exception attempting to read from {1}.',
         ]
-        tty.error(''.join(err_msg).format(spec.short_spec, file_path))
+        tty.error(''.join(err_msg).format(
+            spec.short_spec,
+            specfile_path if specfile_is_json else deprecated_specfile_path))
         tty.debug(url_err)
         tty.warn(result_of_error)
         return rebuild_on_errors
 
-    if not yaml_contents:
-        tty.error('Reading {0} returned nothing'.format(file_path))
+    if not spec_file_contents:
+        tty.error('Reading {0} returned nothing'.format(
+            specfile_path if specfile_is_json else deprecated_specfile_path))
         tty.warn(result_of_error)
         return rebuild_on_errors
 
-    spec_yaml = syaml.load(yaml_contents)
+    spec_dict = (sjson.load(spec_file_contents)
+                 if specfile_is_json else syaml.load(spec_file_contents))
 
-    yaml_spec = spec_yaml['spec']
+    nodes = spec_dict['spec']['nodes']
```

Comment:
This looks like it isn't backwards compatible -- there won't be a `nodes` key if it's the old format.

Maybe
```suggestion
    try:
        nodes = spec_dict['spec']['nodes']
    except KeyError:
        # Prior node dict format omitted 'nodes' key
        nodes = spec_dict['spec']
```